README.md
<p align="center">
	<h1 align="center">lballoons</h2>
	<p align="center">A terminal "screensaver" written in C</p>
</p>
<p align="center">
	<a href="./LICENSE">
		<img alt="License" src="https://img.shields.io/badge/license-GPL-blue?color=7aca00"/>
	</a>
	<a href="https://github.com/LordOfTrident/lballoons/issues">
		<img alt="Issues" src="https://img.shields.io/github/issues/LordOfTrident/lballoons?color=0088ff"/>
	</a>
	<a href="https://github.com/LordOfTrident/lballoons/pulls">
		<img alt="GitHub pull requests" src="https://img.shields.io/github/issues-pr/LordOfTrident/lballoons?color=0088ff"/>
	</a>
	<br><br><br>
	<img width="500px" src="assets/img.png"/>
</p>

I made this purely to practice C since i have written almost no programs in it.
Inspired by [cmatrix](https://github.com/abishekvashok/cmatrix).

## Controls
| Key           | Description      |
|---------------|------------------|
| CTRL+Q/CTRL+C | Quit the program |

## Dependencies
- ncurses

## Make
Use `make all` to see all the make targets.

### Compiling
Run `make` or `make compile` which will create a binary in the `bin/` folder.

> If you want to change the compiler make uses, set the `C` variable. Example: `make C=gcc`


changelog.md
# Changelog
All the versions and the changes they bring

## LBalloons
- `1.0.0`: Working animated balloons
- `1.0.1`: Added colors
- `1.1.0`: Fixed gaps between ballon groups
- `1.1.1`: Bug fix
- `1.1.2`: Working getch
- `1.2.0`: Resizable screen
- `1.2.1`: README.md fixes
- `1.2.2`: Screenshot, dependencies in README.md + new thumbnail
- `1.2.3`: Simple makefile output fix, added void to no-parameters functions
- `1.2.4`: README.md update
- `1.2.5`: Background color changed to default instead of black


src/main.c
/*
 *  lballoons - A terminal "screensaver" written in C
 *  Version 1.2.5
 *  Github: https://github.com/LordOfTrident/lballoons
 *
 *  This is the "art" source code and the one used
 *  by makefile, shaped into a hand holding a balloon.
 *  If you want to see the formatted readable one, look
 *  at readable.c
 */

#include <stdlib.h> // malloc, realloc, free, rand
#include <time.h> // usleep
#include <ncurses.h> // Terminal i/o

#define LB_VERSION_MAJOR 1
#define LB_VERSION_MINOR 2
#define LB_VERSION_PATCH 5

                     //lballoons
                int usleep(unsigned u);
             typedef struct{int x,y;char c
         ;}b;typedef struct{char h,c;}p;const
       char of[6]={1,0,0,1,1,1};const char br[6]
     [4]={"_","/#\\","\\_/","\\","/","\\"};b*o;p*
    r;int a,c,d,e,f,s,l=0,g,m;void n(){int x=0,y=0
  ;for(int i=0;i<s;++i){o[i].x=x*6+rand()%4;o[i].y=
  y*7+rand()%4;o[i].c=rand()%7+1;if(x>=c){x=0;++y;}\
 else++x;}}void t(){o=(b*)malloc(sizeof(b)*s);n();}//
void u(){r=(p*)malloc(sizeof(p)*l);for(int i=0;i<l;++i
){r[i].h=32;r[i].c=1;}}void v(){a=getmaxx(stdscr);c=a/
6;m=getmaxy(stdscr);d=m/7;f=d+2;e=7*(d+1);s=f*(c+1);g=
l;l=a*m;}int main(void){initscr();noecho();start_color
();use_default_colors();curs_set(0);timeout(0);raw();\
cbreak();init_pair(1,7,-1);init_pair(2,4,-1);init_pair
(3,1,-1);init_pair(4,2,-1);init_pair(5,3,-1);init_pair
 (6,6,-1);init_pair(7,5,-1);v();t();u();int z=1;while(
 z){int x=0;for(int i=0;i<s;++i){--o[i].y;if(o[i].y<-
  6){o[i].x=x*6+rand()%4;o[i].y=e;o[i].c=rand()%7+1;
   }b*_=&o[i];for(int i=0;i<6;++i){if((_->y+i)<0||(
    _->y+i)>=m)continue;for(int j=0;br[i][j]!=0&&j
     <4;++j){int p=of[i]+j;if((_->x+p)>=a||(_->x+
      p)<0)break;int pos=(_->y+i)*a+_->x+p;r[pos
        ].h=br[i][j];if(i<3)r[pos].c=_->c;}}if(
          x>=c)x=0;else++x;}move(0,0);for(int
            i=0;i<l;++i){attron(COLOR_PAIR(
             r[i].c));addch(r[i].h);attroff
                (COLOR_PAIR(r[i].c));r\
                   [i].h=32;r[i].c=1
                      ;}refresh();
                          int
                         in=//
                         getch
                           ();
                          switch
                           (in)
                           {case
                            410
                            :v();
                            void
                            *tmp
                            =
                           realloc
                           (o,
                           sizeof
                            (b)*
                             s);
                              if(
                              tmp
                               ==
                               NULL
                               ){
                              free
                              (o);
                              t()
                             ;}
                            else
                            {o=
                            (b*)
                            tmp
                            ;n(
                             );}
                             tmp=
                           realloc
                     (r,sizeof(p)*a*m);\
                   if(tmp==NULL){free//23r3t//3t3f
                  (r);u();}else{r=/*[]'\/.'*///232r33
                    (p*)tmp;if(g<l) {for//rw3rwd--{*9}
                  (int i=g;i<l;++i)/*!@#$%^3rw*///4tawd
                    {r[i].h=32;r[i].c=1;//3428*][--'']
                   }}}break;case'q'&31://2343t38*(64)
                     z=0;}usleep(100000);}/**/\
                   free/*abcdefghijklm_2-*/
                    /*nopqrstuvxyz123*/
                            (o);
                            free
                             (r);
                             endwin
                              ();}


src/readable.c
/*
 *  lballoons - A terminal "screensaver" written in C
 *  Version 1.2.5
 *  Github: https://github.com/LordOfTrident/lballoons
 *
 *  This is the formatted source code NOT used by the
 *  makefile. If you want to see the "art" one, look
 *  at main.c
 */

#include <stdlib.h> // malloc, realloc, free, rand
#include <time.h> // usleep
#include <ncurses.h> // Terminal i/o
#include <stdbool.h> // bool, true, false

#define LB_VERSION_MAJOR 1
#define LB_VERSION_MINOR 2
#define LB_VERSION_PATCH 5

#define COLOR_BACKGROUND -1

int usleep(unsigned u);

typedef struct {
	int x, y;
	char color;
} Balloon;

typedef struct {
	char ch, color;
} Pixel;

// Offset for the balloon lines so i
// dont need to draw spaces which may
// overlay other balloons and cut them
// off
const char offset[6] = {1, 0, 0, 1, 1, 1};
const char balloon_ascii[6][4] = {
	"_",
	"/#\\",
	"\\_/",
	"\\",
	"/",
	"\\"
};

Pixel *screen;
int screen_SizeX;
int screen_SizeY;
int screen_Size1D = 0;
int screen_Size1DPrevious;

Balloon *balloons;
int balloons_SizeX;
int balloons_SizeY;
int balloons_Size1D;
int balloons_SpawnPosition;

bool running;

void screen_WriteAt(
	const char *p_text,
	int p_posX,
	int p_posY,
	char p_color
) {
	if (p_posY < 0 || p_posY >= screen_SizeY)
		return;

	int pos = p_posY * screen_SizeX + p_posX;

	for (int i = 0; p_text[i] != 0; ++ i) {
		if (
			(p_posX + i) >= screen_SizeX ||
			(p_posX + i) < 0
		)
			break;

		screen[pos + i].ch = p_text[i];
		screen[pos + i].color = p_color;
	};
};

void balloon_Draw(const Balloon *p_balloon) {
	for (int i = 0; i < 6; ++ i) {
		screen_WriteAt(
			balloon_ascii[i],
			p_balloon->x + offset[i],
			p_balloon->y + i,
			i < 3? p_balloon->color : 1 // Color the balloon,
			                            // not the rope
		);
	};
};

void ballons_SetDefaults(void) {
	int x = 0, y = 0;
	for (int i = 0; i < balloons_Size1D; ++ i) {
		balloons[i].x = x * 6 + rand() % 4; // Offset the balloon position
		balloons[i].y = y * 7 + rand() % 4;
		balloons[i].color = rand() % 7 + 1;

		if (x >= balloons_SizeX) {
			x = 0;
			++ y;
		} else
			++ x;
	};
};

void balloons_Init(void) {
	balloons = (Balloon*)malloc(sizeof(Balloon) * balloons_Size1D);

	ballons_SetDefaults();
};

void screen_Init(void) {
	screen = (Pixel*)malloc(sizeof(Pixel) * screen_Size1D);

	for (int i = 0; i < screen_Size1D; ++ i) {
		screen[i].ch = ' ';
		screen[i].color = 1;
	};
};

void balloons_Refresh(void) {
	// I had alot of pain when using 2D arrays
	// and realloc, so it led to me switching
	// to 1D arrays which also turned out to
	// be a much better and easier fit
	void *temp = realloc(
		balloons,
		sizeof(Balloon) * balloons_Size1D
	);

	if (temp == NULL) {
		free(balloons);
		balloons_Init();
	} else {
		balloons = (Balloon*)temp;

		ballons_SetDefaults();
	};
};

void screen_Refresh(void) {
	void *temp = realloc(
		screen,
		sizeof(Pixel) * screen_Size1D
	);

	if (temp == NULL) {
		free(screen);
		screen_Init();
	} else {
		screen = (Pixel*)temp;

		if (screen_Size1DPrevious < screen_Size1D) {
			for (
				int i = screen_Size1DPrevious;
				i < screen_Size1D;
				++ i
			) {
				screen[i].ch = ' ';
				screen[i].color = 1;
			};
		};
	};
};

void screen_SetSize(void) {
	screen_SizeX = getmaxx(stdscr);
	screen_SizeY = getmaxy(stdscr);

	screen_Size1DPrevious = screen_Size1D;
	screen_Size1D = screen_SizeX * screen_SizeY;
};

void balloons_SetSize(void) {
	balloons_SizeX = screen_SizeX / 6;
	int balloons_TempSizeY = screen_SizeY / 7;
	balloons_SizeY = balloons_TempSizeY + 2;

	balloons_SpawnPosition = 7 * (balloons_TempSizeY + 1);
	balloons_Size1D = balloons_SizeY * (balloons_SizeX + 1);
};

void Init(void) {
	initscr();

	noecho();
	start_color();
	use_default_colors();
	curs_set(0); // Invisible cursor

	timeout(0); // Make getch not pause the program
	raw();
	cbreak(); // CTRL+C will exit the program, but
	          // the coded-in way is CTRL+Q

	init_pair(1, COLOR_WHITE,   COLOR_BACKGROUND);
	init_pair(2, COLOR_BLUE,    COLOR_BACKGROUND);
	init_pair(3, COLOR_RED,     COLOR_BACKGROUND);
	init_pair(4, COLOR_GREEN,   COLOR_BACKGROUND);
	init_pair(5, COLOR_YELLOW,  COLOR_BACKGROUND);
	init_pair(6, COLOR_CYAN,    COLOR_BACKGROUND);
	init_pair(7, COLOR_MAGENTA, COLOR_BACKGROUND);

	screen_SetSize();
	screen_Init();

	balloons_SetSize();
	balloons_Init();
};

void Finish(void) {
	free(balloons);
	free(screen);

	endwin();
};

void Render(void) {
	move(0, 0);
	for (int i = 0; i < screen_Size1D; ++ i) {
		attron(COLOR_PAIR(screen[i].color));
		addch(screen[i].ch);
		attroff(COLOR_PAIR(screen[i].color));

		screen[i].ch = ' ';
		screen[i].color = 1;
	};

	refresh();
};

void Events(void) {
	int in = wgetch(stdscr);
	switch (in) {
	case KEY_RESIZE:
		screen_SetSize();
		screen_Refresh();

		balloons_SetSize();
		balloons_Refresh();

		break;

	case 'q' & 31:
		running = false;

		break;
	};
};

void MainLoop(void) {
	running = true;
	while (running) {
		int x = 0;
		for (int i = 0; i < balloons_Size1D; ++ i) {
			-- balloons[i].y;

			if (balloons[i].y < -6) {
				balloons[i].x = x * 6 + rand() % 4; // Re-randomize the X offset,
				                                    // not the Y offset
				balloons[i].y = balloons_SpawnPosition;
				balloons[i].color = rand() % 7 + 1;
			};

			balloon_Draw(&balloons[i]);

			if (x >= balloons_SizeX)
				x = 0;
			else
				++ x;
		};

		Render();
		Events();

		usleep(100000);
	};
};

int main(void) {
	Init();
	MainLoop();
	Finish();

	return 0;
};
